File: src/lib/game/state.ts
import { Team } from '$lib';
import { GameEnding, Side, type Square } from '$lib/game/logic';

export type BoardState = {
	teamToPlay: Team,
	castling: CastlingState,
	ending: GameEnding | null,
	enPassantTargetSquare: Square | undefined
}

export type CastlingState = {
	whiteKingSide: boolean,
  whiteQueenSide: boolean,
  blackKingSide: boolean,
  blackQueenSide: boolean,
}

export function updateCastlingState(state: BoardState, team: Team, side: Side, value: boolean) {
	const { castling } = state;

	switch (team) {
		case Team.White:
			state.castling = { ...castling, [side === Side.KINGSIDE ? 'whiteKingSide' : 'whiteQueenSide']: value };
			break;
		case Team.Black:
			state.castling = { ...castling, [side === Side.KINGSIDE ? 'blackKingSide' : 'blackQueenSide']: value };
			break;
		default:
			throw new Error('Invalid team specified');
	}
}

export function cloneState(state: BoardState): BoardState {
	return {
    teamToPlay: state.teamToPlay,
		ending: state.ending,
		enPassantTargetSquare: state.enPassantTargetSquare,
    castling: {
      whiteKingSide: state.castling.whiteKingSide,
      whiteQueenSide: state.castling.whiteQueenSide,
      blackKingSide: state.castling.blackKingSide,
      blackQueenSide: state.castling.blackQueenSide,
    }
  }
}

File: server/src/board/state.rs
use crate::board::{Team, Teams};

#[derive(Debug, Clone, Hash, PartialEq, Eq, Copy)]
pub struct ChessState {
    pub castling_rights: CastlingRights,
    pub en_passant_square: Option<usize>,
    pub team_to_play: Team
}

pub struct CastlingSides;
impl CastlingSides {
    pub const KINGSIDE: usize = 0;
    pub const QUEENSIDE: usize = 1;
}

#[derive(Debug, Clone, Hash, PartialEq, Eq, Copy)]
pub struct CastlingRights(u8);

impl CastlingRights {
    const NONE: u8 = 0b0000;
    const WHITE_KINGSIDE: u8 = 0b0001;
    const WHITE_QUEENSIDE: u8 = 0b0010;
    const BLACK_KINGSIDE: u8 = 0b0100;
    const BLACK_QUEENSIDE: u8 = 0b1000;
    const ALL: u8 = Self::WHITE_KINGSIDE | Self::WHITE_QUEENSIDE | Self::BLACK_KINGSIDE | Self::BLACK_QUEENSIDE;

    pub fn none() -> Self {
        CastlingRights(CastlingRights::NONE)
    }
    pub fn all() -> Self {
        CastlingRights(CastlingRights::ALL)
    }

    pub fn allow(&mut self, team: Team, side: usize) {
        self.0 |= self.get_bit(team, side);
    }

    pub fn disallow(&mut self, team: Team, side: usize) {
        self.0 &= !self.get_bit(team, side);
    }

    pub fn is_allowed(&self, team: Team, side: usize) -> bool {
        self.0 & self.get_bit(team, side) != 0
    }

    pub fn disallow_all(&mut self, team: Team) {
        let bits = match team {
            Teams::WHITE => Self::WHITE_KINGSIDE | Self::WHITE_KINGSIDE,
            Teams::BLACK => Self::BLACK_KINGSIDE | Self::BLACK_QUEENSIDE,
            _ => panic!("Invalid team")
        };
        self.0 &=  !bits;
    }

    pub fn get_bit(&self, team: Team, side: usize) -> u8 {
        match (team, side) {
            (Teams::WHITE, CastlingSides::KINGSIDE) => Self::WHITE_KINGSIDE,
            (Teams::WHITE, CastlingSides::QUEENSIDE) => Self::WHITE_QUEENSIDE,
            (Teams::BLACK, CastlingSides::KINGSIDE) => Self::BLACK_KINGSIDE,
            (Teams::BLACK, CastlingSides::QUEENSIDE) => Self::BLACK_QUEENSIDE,
            _ => 0,
        }
    }
}


